### DISCUSSION TITLE
Custom Button in Property panel

### DISCUSSION THREAD

**Discussion by NourEldin-Ali** [imixs/open-bpmn#253] [General]
Hi @rsoika,

Thank you again for this great project.

I have a question maybe you can help me.
![Screenshot (75)](https://github.com/imixs/open-bpmn/assets/61513661/6b18b215-47cd-47ab-a765-e5083bcb2950)

In your project you add different button to the panel like "add to Data objects", "delete", ...
How I can add a new button and define an action of my button?

Thank you in advance for you reply.
Ali

**Comment by rsoika** [comment]
Hi @NourEldin-Ali 
this control is part of the [JSONForms components](https://jsonforms.io/). This is a UI library based on React to render simple or complex input widgets: https://jsonforms.io/

The widget that you mention is a so called array-renderer.

We are using this concept of JSONForms in the [Property-Panel](https://www.open-bpmn.org/glsp-client/BPMN_PROPERTIES.html) of Open-BPMN. 

If you want to customize this, you usually need to implement a so called custom renderer. We have done this in Open-BPMN also to solve some more complex situations like the new File-Link Component.
You can find the sources here: 
https://github.com/imixs/open-bpmn/tree/master/open-bpmn.glsp-client/open-bpmn-properties/src

But this topic is not easy to answer in general and can be very specific to your use-case. What exactly are you trying to achieve?

**Reply by NourEldin-Ali** [reply]
@rsoika  thank you for you reply.

I need to export a specific BPMN file to change it to other format.
In my thesis, I proposed a new BPMN extenstion, I need to export this extension to other format.

What do you think? what is the right way to do that?
add export button in the panel? or on right click? and could you please refere something can I based on to start?

Thank you.

**Comment by rsoika** [comment]
In Open-BPMN we implemented a [Extension Mechanisim](https://www.open-bpmn.org/glsp-server/BPMN_EXTENSIONS.html). 
This allows you do implement custom extensions for specific parts of your model.
When the `BPMNGModelFactory` class is loading a model it calls the method `applyBPMNExtensions` for each element which calls the method:

` boolean handlesBPMNElement(final BPMNElement bpmnElement);`


This means you can change specific elements of your core model by implementing your own business logic in a BPMNExtension class. So for example you can extend your model with new attributes or even additional elements. 
When the user is next saving the model you got a new 'extended' model.

But if you want to generate separate files when a model is saved, this would be a different scenario. I have not thought about such things. But maybe this could be new idea for OpenBPMN. 

A Kind of a BPMNStorageExtension

```java
public interface BPMNStorageExtension {
    public void onLoad(BPMNModel model, java.io.File source);
    public void onSave(BPMNModel model, java.io.File target);
}
```
Which such a extension you would be able to change the model on load and on save events. For example load additional content from external files or store additional content....

**Comment by rsoika** [comment]
See #254

**Comment by NourEldin-Ali** [comment]
ok thanks
